Feature/upload files assignments and portfolios

wspurgin opened the pull request.

Commit:
Update H2 database to add back course and user FKs

Commit:
Clean up deprecation of AmazonS3Client

Additionally clean up poor variable name warnings, and add error
handling when attempting to list S3 objects.

Commit:
Store File Information for each uploaded file

Commit:
Remove edit icon

Commit:
Fix uploading portfolio items

Commit:
Secure Portfolio Item routes

Commit:
Fix upload by other users to portfolio/assignment

Now the user who uploaded the files is tracked in as the
`uploadedByUserId`, and the "owner" of the document is declared by the
`userId` which will be either a Student or an Instructor.

Commit:
Fix grid spacing when user can't delete

Commit:
Fix errors when 403 and 401 errors.

Was previously always raising a "Token has expired" exception even if
the error wasn't a 401.

Commit:
Use Forbidden status code not Unauthorized

Commit:
Fix students being able to upload to portfolio

Students are currently not allowed to upload to their portfolio. We will
add back as needed.

Commit:
Refactor FileInformation belongs to associations

Directly associate a file information entity with a portfolio or
assignment.

Commit:
Restrict access to Assignments and Uploads

Gracefully deal with Forbidden (403) errors on the frontend

Commit:
Add backend support for FileInformation Resource

This will be used instead of the raw S3 dump that the current interfaces
are using.

Commit:
WIP Assignment and Portfolio file grid rework

Assignments are complete, but Portfolio items still need backend support
of the FileInformation entity.

Commit:
Support FileInformation on PortfolioItem backend

Use exceptions to bubble 404 not found errors instead of returning
`null` in many places on Assignment and Portfolio Upload /
FileInformation actions.

That pattern should be used elsewhere either at the controller or
service level.

Commit:
Fix not adding a create data

Commit:
Add support for File Information in Portfolio

Commit:
Fix response of Store to use DTO

Includes minor cleanup of imports on Portfolio and Assignment Resources.
Additionally ensures that one cannot request an assignment they do have
access to retrieve but fill in a file id that they *don't* have access
to see/edit/delete.

Commit:
Support deleting portfolio files along with item

Commit:
Fix add/delete of file to assignment/portfolio

Commit:
Add delete functionality to portfolio modal

Commit:
Fix delete of file info on assignment deletion

Also deletes data from S3

wspurgin commented:
@michaeledavis-credera 

Can you start reviewing the work thus far? It's a ton of changes, and I kinda want to separate out the work that's remaining because its all front-end. The idea of a Student "submission" of a file to an assignment isn't really flushed out on the front end, but the backend does support it.

I'm just worried about creep here. Any feedback would be 💯

michaeledavis-credera commented:
Sure, I'll start taking a look.

michaeledavis-credera commented on services/src/main/java/org/openlearn/service/AssignmentService.java:
```diff
@@ -147,37 +153,52 @@ public void delete(final Long id) {
 		boolean instructorCheck = true;
 		if (SecurityUtils.isCurrentUserInRole(AuthoritiesConstants.INSTRUCTOR)) {
 			Course course = assignment.getCourse();
 			instructorCheck = user.getId() == course.getInstructor().getId();
```
I know you didn't write this, but this needs to be `user.getId().equals(course.getInstructor().get()` because they are objects, not primitives.

michaeledavis-credera commented on services/src/main/java/org/openlearn/service/PortfolioItemService.java:
```diff
@@ -130,21 +138,26 @@ public PortfolioItemDTO findOne(final Long id) {
 		log.debug("Request to delete portfolio item : {}", id);
 		PortfolioItem portfolioItem = portfolioItemRepository.findOne(id);
-		if (portfolioItem != null && (SecurityUtils.isAdmin() || inOrgOfCurrentUser(portfolioItem))) {
+		if (portfolioItem == null) new PortfolioItemNotFoundException(id);
```
This exception isn't being thrown.

wspurgin replied:
🙈 yikes

michaeledavis-credera commented on services/src/main/java/org/openlearn/service/FileInformationService.java:
```diff
@@ -0,0 +1,129 @@
+	 *
+	 * @param portfolioItem the portfolio item whose files should be deleted.
+	 */
+	public void deleteByPortfolioItem(PortfolioItem portfolioItem) {
```
Shouldn't this be transactional?

wspurgin replied:
All the other services don't have the transactional annotation for their delete methods,  but they probably should be. Especially for these since they delete several different entities.

michaeledavis-credera commented on services/src/main/java/org/openlearn/service/FileInformationService.java:
```diff
@@ -0,0 +1,129 @@
+	 *
+	 * @param assignment the assignment whose files should be deleted.
+	 */
+	public void deleteByAssignment(Assignment assignment) {
```
Shouldn't this be transactional?

wspurgin replied:
Same as above

michaeledavis-credera commented on services/src/main/java/org/openlearn/service/FileInformationService.java:
```diff
@@ -0,0 +1,129 @@
+		return fileInformationTransformer.transform(fileRepository.findOne(id));
+	}
+
+	@Transactional(readOnly = true)
```
Missing JavaDoc

michaeledavis-credera commented on services/src/main/java/org/openlearn/service/FileInformationService.java:
```diff
@@ -0,0 +1,129 @@
+	}
+
+	@Transactional(readOnly = true)
+	public String getFileNameFor(final Long id) {
```
Why is this returning an empty string vs null or throwing an exception?

wspurgin replied:
Didn't want to return `null` to avoid null pointer exception, and I wrote this before I had made the exception class. It should raise the exception though.

michaeledavis-credera commented on services/src/main/java/org/openlearn/service/FileInformationService.java:
```diff
@@ -0,0 +1,129 @@
+	}
+
+	@Transactional(readOnly = true)
+	public String getFileNameFor(final Long id) {
```
Shouldn't this be getFileNameForId to match other methods in this class?

wspurgin replied:
Yes it should

wspurgin replied:
All the other methods in the class don't end in `Id`.

michaeledavis-credera commented on services/src/main/java/org/openlearn/service/PortfolioItemService.java:
```diff
@@ -130,21 +138,26 @@ public PortfolioItemDTO findOne(final Long id) {
-		if (portfolioItem != null && (SecurityUtils.isAdmin() || inOrgOfCurrentUser(portfolioItem))) {
+		if (portfolioItem == null) new PortfolioItemNotFoundException(id);
+
+		if ((SecurityUtils.isAdmin() || inOrgOfCurrentUser(portfolioItem))) {
```
Does this allow students in the same org to delete other students' files?

wspurgin replied:
No right now the `inOrgOfCurrentUser` method enforces that the user has the `ORG_ADMIN` role.

michaeledavis-credera commented on services/src/main/java/org/openlearn/service/PortfolioItemService.java:
```diff
@@ -130,21 +138,26 @@ public PortfolioItemDTO findOne(final Long id) {
 			portfolioItemRepository.delete(id);
 		} else {
-			// TODO: Error handling / logging
+			throw new AccessDeniedException();
```
I'd like to see logging along with these Exceptions.

michaeledavis-credera commented on services/src/main/java/org/openlearn/service/PortfolioItemService.java:
```diff
@@ -130,21 +138,26 @@ public PortfolioItemDTO findOne(final Long id) {
 		User user = userService.getCurrentUser();
 		User student = userRepository.findOneByIdAndAuthority(portfolioItemDTO.getStudentId(), STUDENT);
-		return student != null && user.getOrganization().equals(student.getOrganization());
+		return student != null && SecurityUtils.isCurrentUserInRole(AuthoritiesConstants.ORG_ADMIN) &&
```
This method is called `inOrgOfCurrentUser`, but it looks like it's checking the current user is an **Org Admin** in the portfolio item's organization.

wspurgin replied:
Yeah, I think the permission structure will change here anyway. At the moment that was added to avoid allowing instructors or students in the same organization from having access to actions (like `delete`) that they shouldn't. However, it's all very messy and repetitive. I'm thinking of abstracting the permissions elsewhere into a similar sort of design in Ruby on Rails most popular authorization library [`cancancan`](https://github.com/CanCanCommunity/cancancan). That design is very DRY, and reads well.

michaeledavis-credera commented on services/src/main/java/org/openlearn/service/PortfolioItemService.java:
```diff
@@ -130,21 +138,26 @@ public PortfolioItemDTO findOne(final Long id) {
+	public boolean inOrgOfCurrentUser(final PortfolioItem portfolioItem) {
 		User user = userService.getCurrentUser();
-		return user.getOrganization().equals(portfolioItem.getStudent().getOrganization());
+		return SecurityUtils.isCurrentUserInRole(AuthoritiesConstants.ORG_ADMIN) &&
```
same as above

michaeledavis-credera commented on services/src/main/java/org/openlearn/service/StorageService.java:
```diff
@@ -51,12 +76,13 @@ public StorageService(final FileRepository fileRepository,
 
-		AmazonS3 s3client = new AmazonS3Client(new ProfileCredentialsProvider());
+		User user = userService.getCurrentUser();
+		AmazonS3 s3client = AmazonS3ClientBuilder.defaultClient();
```
Does this need to be reconfigured every time we upload a file?

wspurgin replied:
You mean as opposed to creating a final client that's reused every time? Good question. The builder isn't thread safe, but the `AmazonS3` client **is**. Assuming the client could be created on the main app thread, and then the client shared across multiple server threads, it should be fine. Do you have any concerns with the client being built every time?

michaeledavis-credera commented on services/src/main/java/org/openlearn/service/StorageService.java:
```diff
@@ -68,39 +94,48 @@ public FileInformation store(final MultipartFile file, Long assignmentId, Long p
 		}
 
 
-//		Course course = new Course();//courseRepository.findOne(courseId);
```
It concerns me that the try/catch above is just swallowing exceptions.

wspurgin replied:
Yup. That's a holdover from the first implementation of file uploads. It caused problems for me earlier when I was accidentally using Taylor's bucket and had the store process returning successfully even though I couldn't actually upload to that bucket. I'm in the works of refactoring that catch to log and throw again (and only catch amazon exceptions).

michaeledavis-credera commented on services/src/main/java/org/openlearn/service/StorageService.java:
```diff
@@ -68,39 +94,48 @@ public FileInformation store(final MultipartFile file, Long assignmentId, Long p
+		fileInformation.setFileUrl("https://s3.amazonaws.com/" + uploadBucketName + "/" + keyName);
+		fileInformation.setUploadedByUser(user);
+		fileInformation.setCreatedDate(ZonedDateTime.now());
+		if (assignmentId != null) {
```
We aren't validating that either the assignmentId or portfolioId are there.

wspurgin replied:
Yeah I have that in the works. Validation is a slow moving beast because I've been adding it in a lot of places

michaeledavis-credera commented on services/src/main/java/org/openlearn/service/StorageService.java:
```diff
@@ -51,12 +76,13 @@ public StorageService(final FileRepository fileRepository,
 	 */
 	//TODO cbernal fix blue documentation
-	public FileInformation store(final MultipartFile file, Long assignmentId, Long portfolioId) {
+	public FileInformationDTO store(final MultipartFile file, Long assignmentId, Long portfolioId) {
```
Should this method be transactional?

wspurgin replied:
Perhaps. Taylor didn't make it transactional initially, but I don't know his reasoning. My only guess is that since storing a file *can* take some time (depending on the size), it might be because we didn't want to hog resources with a long running transaction.

wspurgin replied:
I'd have to look into how H2 implements transactions to have any clear indications of what tradeoffs there might be.

michaeledavis-credera commented on services/src/main/java/org/openlearn/service/StorageService.java:
```diff
@@ -68,39 +94,48 @@ public FileInformation store(final MultipartFile file, Long assignmentId, Long p
+		return fileInformationTransformer.transform(fileRepository.save(fileInformation));
 	}
 
 	private static File convertToFile(MultipartFile file) throws IOException {
```
It seems strange that we have to write this file to disk to convert it.

wspurgin replied:
Agreed. Moreover, I don't think that file is every deleted once its converted.

wspurgin replied:
Again a holdover from the first implementation, but I'll look into fixing it.

michaeledavis-credera commented on services/src/main/java/org/openlearn/service/StorageService.java:
```diff
@@ -109,23 +144,101 @@ public InputStream getUpload(Long assignmentId, Long portfolioId, String fileNam
+			s3client.deleteObject(new DeleteObjectRequest(bucket, key));
+			fileRepository.delete(fileInformation);
 		} catch(Exception e) {
 			log.error(e.getMessage());
```
This exception is getting swallowed.

wspurgin replied:
Yeah, holdover from the first implementation, it's in the works of being fixed.

michaeledavis-credera commented on services/src/main/java/org/openlearn/service/StorageService.java:
```diff
@@ -109,23 +144,101 @@ public InputStream getUpload(Long assignmentId, Long portfolioId, String fileNam
 	}
 
-	private String getS3FilePrefix(Long assignmentId, Long portfolioId) {
+	public void deleteUploads(List<FileInformation> files) {
```
This method is pretty intense. Let's see if we can refactor it so that it's more clear and readable.

wspurgin replied:
What's your recommendation there? We can separate it into multiple calls to private methods? That would at least make the purpose clear. Is that along your line of thinking?

michaeledavis-credera commented on services/src/main/java/org/openlearn/service/StorageService.java:
```diff
@@ -109,23 +144,101 @@ public InputStream getUpload(Long assignmentId, Long portfolioId, String fileNam
+		return bucketAndKey.substring(separatingIndex+1);
+	}
+
+	private String retrieveFullPath(FileInformation fileInformation) {
```
It seems like there is some room to improve how we handle buckets and keys. Can we not just store the bucket and key in the FileInformation?

wspurgin replied:
We could, but Normalization purist would claim that the bucket and key are "calculable" from the file url. I'm also of a mind that storing the bucket and key would be ideal, the only problem is the handling of existing data. How do migrate existing schema and data correctly? I don't think that's too big of an issue (since it's just our development environments), but what about the staging environment we've been deploying to? Don't know if that DB is wiped every time...

michaeledavis-credera commented on services/src/main/java/org/openlearn/web/rest/AssignmentResource.java:
```diff
@@ -101,8 +108,12 @@ public ResponseEntity getByCourse(@PathVariable final Long id, @ApiParam final P
+			AssignmentDTO response = assignmentService.save(assignmentDTO);
+			return ResponseEntity.created(new URI(ENDPOINT + response.getId())).body(response);
+		} else {
+			return new ResponseEntity(HttpStatus.FORBIDDEN);
```
I'd like to see logging in these failure cases.

michaeledavis-credera commented on services/src/main/java/org/openlearn/web/rest/AssignmentResource.java:
```diff
@@ -146,58 +165,118 @@ public ResponseEntity delete(@PathVariable final Long id) {
-			"You successfully uploaded " + file.getOriginalFilename() + "!");
+		if (canUploadFilesToAssignment(assignmentService.findOne(assignmentId))) {
+			FileInformationDTO response = storageService.store(file, assignmentId, null);
+			redirectAttributes.addFlashAttribute("message",
```
What is this flash attribute for?

wspurgin replied:
Holdover from the original implementation. It just got indented (which is why it looks like an addition). I'm not sure what it does or why it was added to be honest. Doesn't look like this endpoint does any redirect...

michaeledavis-credera commented on services/src/main/java/org/openlearn/web/rest/AssignmentResource.java:
```diff
@@ -146,58 +165,118 @@ public ResponseEntity delete(@PathVariable final Long id) {
-				return new ResponseEntity(out, responseHeaders, HttpStatus.OK);
-			} catch (Exception e) {
-				new ResponseEntity("File Not Found", HttpStatus.NOT_FOUND);
+		if (hasCreateUpdateDeleteAuthority(assignmentDTO) && fileInformationDTO.getAssignmentId() == assignmentId) {
```
This needs to be .equals instead of ==, as they are objects.

michaeledavis-credera commented on services/src/main/java/org/openlearn/web/rest/AssignmentResource.java:
```diff
@@ -146,58 +165,118 @@ public ResponseEntity delete(@PathVariable final Long id) {
+
+					HttpHeaders responseHeaders = new HttpHeaders();
+					responseHeaders.add("content-disposition", "attachment; filename=" + fileName);
+					// responseHeaders.add("Content-Type", type);
```
Remove commented out code.

wspurgin replied:
Another holdover that got indented.

michaeledavis-credera commented on services/src/main/java/org/openlearn/web/rest/AssignmentResource.java:
```diff
@@ -146,58 +165,118 @@ public ResponseEntity delete(@PathVariable final Long id) {
+
+					return new ResponseEntity(out, responseHeaders, HttpStatus.OK);
+				} catch (Exception e) {
+					new ResponseEntity("File Not Found", HttpStatus.NOT_FOUND);
```
This line doesn't do anything.

wspurgin replied:
Nope that's another hold over that got indented.

michaeledavis-credera commented on services/src/main/java/org/openlearn/web/rest/AssignmentResource.java:
```diff
@@ -146,58 +165,118 @@ public ResponseEntity delete(@PathVariable final Long id) {
-			return new ResponseEntity("File Not Found", HttpStatus.NOT_FOUND);
 		} else {
-			return new ResponseEntity ("File Not Found", HttpStatus.NOT_FOUND);
+			return new ResponseEntity(HttpStatus.FORBIDDEN);
```
You might want to consider returning responses/throwing exceptions earlier to avoid all of this if/else nesting.

wspurgin replied:
Agreed. Or letting the service(s) do the throwing. This is mostly legacy code which didn't have any error handling or exception, just returning `null`s so that's why there's all this if else stuff.

michaeledavis-credera commented on services/src/main/java/org/openlearn/web/rest/AssignmentResource.java:
```diff
@@ -146,58 +165,118 @@ public ResponseEntity delete(@PathVariable final Long id) {
+		if (assignmentDTO == null) throw new AssignmentNotFoundException(assignmentId);
+		if (fileInformationDTO == null) throw new FileInformationNotFoundException(id);
+
+		if (hasCreateUpdateDeleteAuthority(assignmentDTO) && fileInformationDTO.getAssignmentId() == assignmentId) {
```
== -> .equals

michaeledavis-credera commented on services/src/main/java/org/openlearn/web/rest/FileResource.java:
```diff
@@ -0,0 +1,22 @@
+
+	private static final String ENDPOINT = "/api/files/";
+
+	private static final Logger log = LoggerFactory.getLogger(CourseResource.class);
```
Use FileResource.class here.

wspurgin replied:
🙈 Copy-pasta error

michaeledavis-credera commented on services/src/main/java/org/openlearn/web/rest/PortfolioItemResource.java:
```diff
@@ -100,8 +113,13 @@ public ResponseEntity getPortfolio(@PathVariable final Long id) {
+			PortfolioItemDTO response = portfolioItemService.save(portfolioItemDTO);
+			return ResponseEntity.created(new URI(ENDPOINT + response.getId())).body(response);
+		} else {
+			return new ResponseEntity(HttpStatus.FORBIDDEN);
```
Error logs

michaeledavis-credera commented on services/src/main/java/org/openlearn/web/rest/PortfolioItemResource.java:
```diff
@@ -129,9 +151,14 @@ public ResponseEntity update(@RequestBody @Valid final PortfolioItemDTO portfoli
-		return ResponseEntity.ok().build();
+		PortfolioItemDTO portfolioItem = portfolioItemService.findOne(id);
+		if (hasCreateUpdateDeleteAuthority(portfolioItem)) {
+			log.debug("DELETE request to delete portfolio item : {}", id);
```
This log should be up top in the method, like the other methods.

michaeledavis-credera commented on services/src/main/java/org/openlearn/web/rest/PortfolioItemResource.java:
```diff
@@ -129,9 +151,14 @@ public ResponseEntity update(@RequestBody @Valid final PortfolioItemDTO portfoli
+		if (hasCreateUpdateDeleteAuthority(portfolioItem)) {
+			log.debug("DELETE request to delete portfolio item : {}", id);
+			portfolioItemService.delete(id);
+			return ResponseEntity.ok().build();
```
200s (OKs) with no body should be 204(NO CONTENT)s

wspurgin replied:
Yup. Again, just a spacing change here (with the `if`). We'll probably just need to do a cleanup PR to handle all that.

wspurgin replied:
Because it's that way in every Resource controller

michaeledavis-credera commented on services/src/main/java/org/openlearn/web/rest/PortfolioItemResource.java:
```diff
@@ -144,57 +171,110 @@ public ResponseEntity delete(@PathVariable final Long id) {
 	@Secured({AuthoritiesConstants.ADMIN, AuthoritiesConstants.ORG_ADMIN, AuthoritiesConstants.INSTRUCTOR, AuthoritiesConstants.STUDENT})
-	public String uploadCourseFile(@PathVariable final Long portfolioId,
+	public ResponseEntity uploadCourseFile(@PathVariable final Long portfolioId,
 								   @RequestParam("file") MultipartFile file,
```
Spaces mixed with tabs.

wspurgin replied:
Ugh. My b. Editor frenzy

michaeledavis-credera commented on services/src/main/java/org/openlearn/web/rest/PortfolioItemResource.java:
```diff
@@ -144,57 +171,110 @@ public ResponseEntity delete(@PathVariable final Long id) {
+		PortfolioItemDTO portfolioItem = portfolioItemService.findOne(portfolioId);
+		if (canUploadFilesToPortfolio(portfolioItem)) {
+			FileInformationDTO response = storageService.store(file, null, portfolioId);
+			redirectAttributes.addFlashAttribute("message",
```
What's the point of this flash attribute?

wspurgin replied:
Don't know, this isn't actually an addition, it's just a spacing change (because of the higher level `if` I added. I'm not sure why this was originally added or what it does.

michaeledavis-credera commented on services/src/main/java/org/openlearn/web/rest/PortfolioItemResource.java:
```diff
@@ -144,57 +171,110 @@ public ResponseEntity delete(@PathVariable final Long id) {
-				return new ResponseEntity(out, responseHeaders, HttpStatus.OK);
-			} catch (Exception e) {
-				new ResponseEntity("File Not Found", HttpStatus.NOT_FOUND);
+		if (canUploadFilesToPortfolio(portfolioItem) && fileInformationDTO.getPortfolioItemId() == portfolioId) {
```
== to .equals

michaeledavis-credera commented on services/src/main/java/org/openlearn/web/rest/PortfolioItemResource.java:
```diff
@@ -144,57 +171,110 @@ public ResponseEntity delete(@PathVariable final Long id) {
+
+					HttpHeaders responseHeaders = new HttpHeaders();
+					responseHeaders.add("content-disposition", "attachment; filename=" + fileName);
+					// responseHeaders.add("Content-Type", type);
```
Remove commented code.

wspurgin replied:
Same thing as with the `AssignmentResource`. Legacy code that got indented.

michaeledavis-credera commented on services/src/main/java/org/openlearn/web/rest/PortfolioItemResource.java:
```diff
@@ -144,57 +171,110 @@ public ResponseEntity delete(@PathVariable final Long id) {
+
+					return new ResponseEntity(out, responseHeaders, HttpStatus.OK);
+				} catch (Exception e) {
+					new ResponseEntity("File Not Found", HttpStatus.NOT_FOUND);
```
This isn't doing anything.

michaeledavis-credera replied:
Also a failure in marshalling the file should result in a 500 rather than a 404.

wspurgin replied:
Again that's a holdover. I think catching the general exception is part of the problem here. I'll refactor this hold methods nested `if` nastiness anyway. Perhaps we should return a [`418`](https://www.google.com/teapot)?

michaeledavis-credera commented on services/src/main/java/org/openlearn/web/rest/PortfolioItemResource.java:
```diff
@@ -144,57 +171,110 @@ public ResponseEntity delete(@PathVariable final Long id) {
+		if (portfolioItem == null) throw new PortfolioItemNotFoundException(portfolioId);
+		if (fileInformationDTO == null) throw new FileInformationNotFoundException(id);
+
+		if (canUploadFilesToPortfolio(portfolioItem) && fileInformationDTO.getPortfolioItemId() == portfolioId) {
```
== to .equals

michaeledavis-credera commented on www/src/app/controls/course/files-grid/files-grid.component.ts:
```diff
@@ -66,58 +69,63 @@ export class FilesGridComponent implements OnInit {
         this.files = files;
         for (let fileIdx = 0; fileIdx < this.files.length; fileIdx++) {
-          files[fileIdx].key = files[fileIdx].key.substring(files[fileIdx].key.indexOf("_", 2)+1);
+          let fileUrl = this.files[fileIdx].fileUrl
```
Prefer const over let whenever possible.

michaeledavis-credera commented on www/src/app/shared/auth/http-wrapper.service.ts:
```diff
@@ -44,11 +44,17 @@ export class HttpWrapperService {
-            this.notify.error('You have been automatically signed out. Please sign in again.')
+          this.logoutService.logout();
+          this.router.navigate(['/login']);
+          this.notify.error('You have been automatically signed out. Please sign in again.')
```
I missed a semicolon. Mind adding that back?

michaeledavis-credera commented on www/src/app/shared/auth/http-wrapper.service.ts:
```diff
@@ -44,11 +44,17 @@ export class HttpWrapperService {
+        } else if (response.status === 403) {
+          // They were forrbidden to do whatever attempted action
+          // this.router.navigate(['/dashboard']);
+          this.notify.error('You are not permitted to perform that action.')
```
semicolon here too.

michaeledavis-credera commented on www/src/app/shared/auth/http-wrapper.service.ts:
```diff
@@ -44,11 +44,17 @@ export class HttpWrapperService {
-            this.logoutService.logout();
-            this.router.navigate(['/login']);
-            this.notify.error('You have been automatically signed out. Please sign in again.')
+          this.logoutService.logout();
```
This whole file is spaced with 4 spaces, but here it's only 2 now.

wspurgin replied:
Technically the whole file should be 2 spaces. That's the convention (as I understand it). Since I was editing the code, my editor indented the block I was in to the right level, but left the rest of the file untouched. I could re-indent the whole file, but that might be a lot of extra change that this PR doesn't need. If it is quite irksome, I'll do it in a subsequent PR. 👍

michaeledavis-credera commented on www/src/app/shared/auth/http-wrapper.service.ts:
```diff
@@ -44,11 +44,17 @@ export class HttpWrapperService {
+          return Observable.throw(new Error('Token has expired. Logging the user out now.'));
+        } else if (response.status === 403) {
+          // They were forrbidden to do whatever attempted action
+          // this.router.navigate(['/dashboard']);
```
Remove commented out code.

michaeledavis-credera commented on www/src/app/shared/file-guardian.service.ts:
```diff
@@ -0,0 +1,25 @@
+      instance.id != null
+  }
+
+  isPortfolio(instance: any): instance is PortfolioItem {
```
I've never seen this `instance is PortfoioItem` syntax. How does this work?

wspurgin replied:
It's a [`type predicate`](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards). Read about it when making this method.

Commit:
Fix using `==` over equals on non-primitives

Commit:
Validate existence of portfolio and assignment

and do so before attempting to upload to S3

Commit:
Clean up and fixes from PR feedback

- Adds logging in Assignment and Portfolio Item resources
- Fixes various stylistic and missing documentation concerns
- Removes several unused imports
- Favors raising errors (that are catchable in the Spring web
    interface) over returning null and having if blocks.
- Fixes areas in the Assignment and Portfolio Resources that were
    swallowing errors.

Commit:
Reimplement upload without creating to local file

Commit:
Add error handling around S3 errors

Commit:
Fix mixed indention of tabs and spaces

Commit:
Don't attempt to reconstruct the file entity

Just create the "key" name and then push to files array

wspurgin commented:
@michaeledavis-credera Rebased on develop. Addressed the feedback to the best of my ability. Some things I think will be more suitable to subsequent PRs. Let me know if you have any other feedback!

Commit:
Add error handling for when a download fails

Commit:
Hide extra column if student view

Don't get *all* student assignments (includes grades) if a student.

Commit:
Fix [Org] Admin uploading on behalf of Instructor

Commit:
Add instructor uploads endpoint

Commit:
Show instructor uploaded files for students

Commit:
Fix comparing user object to id
